src/attribute_context.cc

```diff
@@ -0,0 +1,98 @@
+/* Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "src/attribute_context.h"
+#include "google/protobuf/timestamp.pb.h"
+
+using ::google::protobuf::Map;
+using ::google::protobuf::Timestamp;
+
+namespace istio {
+namespace mixer_client {
+namespace {
+
+// TODO: add code to build context to reduce attributes.
+// Only check these attributes to build context.
+std::set<std::string> kContextSet = {"serviceName", "peerId", "location",
+                                     "apiName", "apiVersion"};
+
+// Convert timestamp from time_point to Timestamp
+Timestamp CreateTimestamp(std::chrono::system_clock::time_point tp) {
+  Timestamp time_stamp;
+  long long nanos = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                        tp.time_since_epoch())
+                        .count();
+
+  time_stamp.set_seconds(nanos / 1000000000);
+  time_stamp.set_nanos(nanos % 1000000000);
+  return time_stamp;
+}
+
+}  // namespace
+
+void AttributeContext::FillProto(const Attributes& attributes,
+                                 ::istio::mixer::v1::Attributes* pb) {
+  // TODO build context use kContextSet to reduce attributes.
+
+  // Fill attributes.
+  int next_dict_index = dict_map_.size();
+  bool dict_changed = false;
+  for (const auto& it : attributes.attributes) {
+    const std::string& name = it.first;
+
+    // Find index for the name.
+    int index;
+    const auto& dict_it = dict_map_.find(name);
+    if (dict_it == dict_map_.end()) {
```

Comment:
Are we fine with growing dict_map_ whenever we see new attribute, but not deleting from it? That means in worst case (you see a new attribute in every request) dictionary could be large. If it is not concerned now, worth a comment?